### PR TITLE
Mzuevx/ac librispeech

### PR DESCRIPTION
**Pull request by mzuevx**
PR implements sound features extractor pipeline code and WER metric code. Both needed for DeepSpeech accuracy test.

### CODE REVIEW

**Comment by mzuevx** [comment]
@eaidova , please review

**Comment by eaidova** [comment]
Jenkins ok to test

File: tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import string
+import numpy as np
+
+from ..adapters import Adapter
+from ..config import ConfigValidator, NumberField, BoolField, ConfigError
+from ..representation import CharacterRecognitionPrediction
+
+
+class CTCBeamSearchDecoder(Adapter):
+    __provider__ = 'ctc_beam_search_decoder'
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py:26]: what the difference between your adapter and my beam_search_decoder?
https://github.com/opencv/open_model_zoo/blob/master/tools/accuracy_checker/accuracy_checker/adapters/text_recognition.py#L25

File: tools/accuracy_checker/accuracy_checker/annotation_converters/librispeech.py [hunk 2]
```diff
@@ -0,0 +1,1614 @@
...
+from .format_converter import ConverterReturn
+from scipy.fftpack import dct
+from scipy.signal import stft
+
+class LibrispeechConverter(DirectoryBasedAnnotationConverter):
+    __provider__ = 'librispeech'
+    annotation_types = (CharacterRecognitionAnnotation, )
+    HZ2MEL = 1127.0
+
+    # @classmethod
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/librispeech.py:33]: please do not forget to remove debug comments

File: tools/accuracy_checker/accuracy_checker/representation/asr_representation.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class ASRAnnotation(ASRRepresentation):
+    def __init__(self, identifier, source='', reference=''):
+        super().__init__(identifier)
+        self.source = source
+        self.reference = reference
+
+
+class ASRPrediction(ASRRepresentation):
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/representation/asr_representation.py:15]: your adapter returns CharacterRecognitionPrediction, this class is not used

File: tools/accuracy_checker/accuracy_checker/representation/asr_representation.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+from .base_representation import BaseRepresentation
+
+
+class ASRRepresentation(BaseRepresentation):
+    pass
+
+
+class ASRAnnotation(ASRRepresentation):
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/representation/asr_representation.py:8]: metric expects CharacterRecognitionAnnotation

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    def configure(self):
+        self.overall_metric = []
+
+    def update(self, annotation, prediction):
+
+        # distance = editDistance(annotation, prediction)
+        h = prediction
+        r = annotation
+
+        dist = numpy.zeros((len(r) + 1) * (len(h) + 1), dtype=numpy.uint8).reshape((len(r) + 1, len(h) + 1))
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py:43]:
```suggestion
        dist = numpy.zeros((len(r) + 1), (len(h) + 1)), dtype=numpy.uint8)
```

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                if r[i - 1] == h[j - 1]:
+                    dist[i][j] = dist[i - 1][j - 1]
+                else:
+                    substitute = dist[i - 1][j - 1] + 1
+                    insert = dist[i][j - 1] + 1
+                    delete = dist[i - 1][j] + 1
+                    dist[i][j] = min(substitute, insert, delete)
+
+        # print the result in aligned way
+        result = float(d[len(r)][len(h)]) / len(r)
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py:59]: what is `d`, I do not see this variable assignment

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    overall_metric = []
+
+    def configure(self):
+        self.overall_metric = []
+
+    def update(self, annotation, prediction):
+
+        # distance = editDistance(annotation, prediction)
+        h = prediction
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py:40]: here something wrong, do you remember that prediction is CharacterRecognitionPrediction class object?
I believe, you mean prediction.label

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+from .metric import PerImageEvaluationMetric
+from ..utils import finalize_metric_result
+
+
+class SpeechRecognitionWER(PerImageEvaluationMetric):
+    __provider__ = 'wer'
+    annotation_types = (CharacterRecognitionAnnotation,)
+    prediction_types = (CharacterRecognitionPrediction,)
+
+    overall_metric = []
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py:32]: no need to specify here if you have it in configure (this method call inside __init__())

File: tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py (hunk 1 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py:26]: I tried your implementation but output result differs from reference. This implementation of beam_search_decoder calculates result that match reference.

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py (hunk 7 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py]: fixed

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py (hunk 6 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py]: It's my mistake, fixed

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py (hunk 5 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py]: fixed

File: tools/accuracy_checker/accuracy_checker/annotation_converters/librispeech.py (hunk 2 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/annotation_converters/librispeech.py]: fixed

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py (hunk 8 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py]: fixed

File: tools/accuracy_checker/accuracy_checker/representation/asr_representation.py (hunk 4 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/representation/asr_representation.py]: fixed, extra class removed

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class SpeechRecognitionWER(PerImageEvaluationMetric):
+    __provider__ = 'wer'
+    annotation_types = (CharacterRecognitionAnnotation,)
+    prediction_types = (CharacterRecognitionPrediction,)
+
+    def configure(self):
+        self.overall_metric = []
+        self.meta = {'target': 'lower-better'}
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py:32]:
```suggestion
        self.meta.update({'target': 'higher-worse'})
```

**Comment by mzuevx** [comment]
@eaidova , I fix all Pylint warning except 2 false positive. Seems that this PR can be merged.

File: tools/accuracy_checker/accuracy_checker/annotation_converters/librispeech.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from pathlib import Path
+import re
+
+from ..representation import CharacterRecognitionAnnotation
+# from ..config import NumberField, StringField, PathField, BoolField
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/annotation_converters/librispeech.py:21]: please remove commented line

File: tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class SpeechRecognitionWER(PerImageEvaluationMetric):
+    __provider__ = 'wer'
+    annotation_types = (CharacterRecognitionAnnotation,)
+    prediction_types = (CharacterRecognitionPrediction,)
+
+    def configure(self):
+        self.overall_metric = []
+        self.meta = {'target': 'higher-worse'}
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/metrics/speech_recognition.py:32]: please use update, super class can introduce some specific keys in meta
```suggestion
        self.meta['target'] = 'higher-worse'
```

File: tools/accuracy_checker/accuracy_checker/preprocessor/audio_preprocessing.py [hunk 12]
```diff
@@ -178,9 +179,307 @@ def _parse_duration(self, duration):
...
+        self.fftbase = self.get_value_from_config('fftbase')
+        self.magnutide_squared = self.get_value_from_config('magnitude_squared')
+        self.skip_channels = self.get_value_from_config('skip_channels')
+
+    def process(self, image, annotation_meta=None):
+        frames = image.data
+        if self.skip_channels:
+            frames = frames.squeeze()
+
+        pspec = np.absolute(np.fft.rfft(frames, self.fftbase))
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/preprocessor/audio_preprocessing.py:255]: ignore pylint false positive
```suggestion
        pspec = np.absolute(np.fft.rfft(frames, self.fftbase)) # pylint:disable=W9904
```

File: tools/accuracy_checker/accuracy_checker/preprocessor/audio_preprocessing.py [hunk 13]
```diff
@@ -178,9 +179,307 @@ def _parse_duration(self, duration):
...
+        self.step = self.get_value_from_config('step')
+
+    def process(self, image, annotation_meta=None):
+        features = image.data
+
+        steps, context, numceps = features.shape
+        if steps % self.step:
+            empty_context = np.zeros((self.step - (steps % self.step), context, numceps), dtype=features.dtype)
+            features = np.concatenate((features, empty_context))
+            steps, context, numceps = features.shape
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/preprocessor/audio_preprocessing.py:474]:
```suggestion
            steps, context, numceps = features.shape # pylint:disable=E0633
```

File: tools/accuracy_checker/accuracy_checker/annotation_converters/librispeech.py (hunk 10 above)
Comment (eaidova) [reply tools/accuracy_checker/accuracy_checker/annotation_converters/librispeech.py:21]:
```suggestion
```

**Pull request merged**